cirq/protocols/apply_mixture_protocol.py

```diff
@@ -0,0 +1,368 @@
+                        "auxiliary_buffer{}. This type violates the contract "
+                        "that appears in apply_mixture's documentation.".format(
+                            type(val), buf_num_str))
+
+            assert result is not args.auxiliary_buffer0, err_str('0')
+            assert result is not args.auxiliary_buffer1, err_str('1')
+            return result
+
+    # Possibly use `cirq.apply_unitary`. (STEP B)
+    result = _apply_unitary_strat(val, args, is_dm)
+    if result is not None:
+        return result
+
+    # Fallback to using the object's `_mixture_` matrices. (STEP C)
+    prob_mix = mixture(val, None)
+    if prob_mix is not None:
+        return _mixture_strat(prob_mix, args, is_dm)
+
+    # Don't know how to apply mixture. Fallback to specified default behavior.
+    # (STEP D)
+    if default is not RaiseTypeErrorIfNotProvided:
+        return default
+    raise TypeError(
+        "object of type '{}' has no _apply_mixture_, _apply_unitary_, "
+        "_unitary_, or _mixture_ methods (or they returned None or "
+        "NotImplemented).".format(type(val)))
+
+
+def _validate_input(val: Any, args: 'ApplyMixtureArgs'
+                   ) -> Tuple[Any, 'ApplyMixtureArgs', bool]:
+    """Validate args input and determine if we are operating on a
+    density matrix or a wavefunction.
+    """
+
+    is_dm = False
+    val_qid_shape = qid_shape_protocol.qid_shape(val,
+                                                 (2,) * len(args.left_axes))
+    left_shape = tuple(args.target_tensor.shape[i] for i in args.left_axes)
+    if val_qid_shape != left_shape:
+        raise ValueError('Invalid mixture qid shape is not equal to the '
+                         'selected left and right shape of target_tensor. '
+                         'Got {!r} but expected {!r}.'.format(
+                             val_qid_shape, left_shape))
+
+    if args.right_axes is not None:
+        is_dm = True
+
+        right_shape = tuple(
+            args.target_tensor.shape[i] for i in args.right_axes)
+        if left_shape != right_shape:
+            raise ValueError(
+                'Invalid target_tensor shape or selected axes. '
+                'The selected left and right shape of '
+                'target_tensor are not equal. Got {!r} and {!r}.'.format(
+                    left_shape, right_shape))
+
+    return val, args, is_dm
+
+
+def _apply_unitary_strat(val: Any, args: 'ApplyMixtureArgs',
+                         is_dm: bool) -> Optional[np.ndarray]:
+    """Attempt to use `apply_unitary` and return the result.
+
+    If `val` does not support `apply_unitary` returns None.
+    """
+    left_args = ApplyUnitaryArgs(target_tensor=args.target_tensor,
+                                 available_buffer=args.auxiliary_buffer0,
+                                 axes=args.left_axes)
+    left_result = apply_unitary(val, left_args, None)
+    if left_result is None:
+        return None
+
+    if not is_dm:
```

Comment:
Expand to `is_density_matrix` everywhere

Reply:
fixed